### PR TITLE
Updated exposure function call: enable temporal filtering

### PR DESCRIPTION
**Pull request by erialC-P**
This PR updates the exposure function to incorporate the ability to include temporal filtering in the exposure output/s. Incorporating this update into the intertidal repo is required to enable moving the exposure demo notebook into the dea_notebooks repository whilst ensuring enabled functionality is provided by the intertidal repository.

Changes and updates:
- all updates to the function call default to `None` so theoretically, no updates should be required to the remainder of our production code base.
- [Important] returned exposure values are now included in a dict object. A working prototype for handling this is included in the `Intertidal_workflow.ipynb`. 
- Optional inclusions in this new exposure function include the ability to apply temporal filters to the exposure high temporal resolution tidal modelling (e.g. winter dates only). Separately, tuples of filters can be applied (e.g. winter daytime). Lastly, functionality is built in to return the tide modelling and filtered tide height values to enable demonstration of the exposure method in other workflows.

Future work:
- combine single and tuple filter options into single function call
- thoroughly check all docstrings to ensure they are up to date

### CODE REVIEW

File: intertidal/elevation.py [hunk 1]
```diff
@@ -1264,22 +1264,24 @@ def intertidal_cli(
...
+                end_date=end_date,
+                modelled_freq=modelled_freq,
                 tide_model=tide_model,
                 tide_model_dir=tide_model_dir,
-                run_id=run_id,
-                log=log,
             )
+            
+            # Write each exposure output as new variables in the main dataset
+            for x in exposure_filters.data_vars:
```
Comment (robbibt) [inline intertidal/elevation.py]: I think given we're only ever going to be extracting the "unfiltered" layer here, I think we can simplify this code to:
```
ds["exposure"] = exposure_filters["unfiltered"]
```

File: intertidal/exposure.py [hunk 2]
```diff
@@ -1,35 +1,286 @@
...
+    that each pixel was 'exposed' from tidal inundation during the time
+    period of interest.
+
     The exposure calculation is based on tide-height differences between
     the elevation value and modelled tide height percentiles.
 
+    For an 'unfiltered', all of epoch-time, analysis, exposure is
+    calculated per pixel. All other filter options calculate exposure
+    from high temporal resolution modelled tides that are averaged 
+    across the nominated area of interest only.
```
Comment (robbibt) [inline intertidal/exposure.py]:
```suggestion
    into a 1D timeseries across the nominated area of interest.
```

File: intertidal/exposure.py [hunk 3]
```diff
@@ -80,31 +389,163 @@ def exposure(
...
-        calculate_quantiles=tide_percentiles,
-        times=times,
+        ancillary_points="data/raw/tide_correlations_2017-2019.geojson",
+        resample=False,
     )
 
-    # Calculate the tide-height difference between the elevation value and
-    # each percentile value per pixel
-    diff = abs(tide_cq - dem)
+    modelledtides_1d = modelledtides_lowres.mean(dim=["x", "y"])
```
Comment (robbibt) [inline intertidal/exposure.py:454]: There's a similar code block below this that is commented out - can you have a look and see if we can remove one of these?

File: intertidal/exposure.py [hunk 4]
```diff
@@ -80,31 +389,163 @@ def exposure(
...
+        modelledtides_x = modelledtides_1d.sel(time=timeranges[str(x)])
+
+        # Calculate quantile values on remaining tide heights
+        modelledtides_x = (
+            modelledtides_x.quantile(q=calculate_quantiles, dim="time")
+            .to_dataset()
+            .tide_m
+        )
+
+        # Add modelledtides_x to output dict
```
Comment (robbibt) [inline intertidal/exposure.py]:
```suggestion
        # Add modelledtides_x to output dataset
```

File: intertidal/exposure.py [hunk 5]
```diff
@@ -80,31 +389,163 @@ def exposure(
...
+
+    # Intersect the filters of interest to extract the common datetimes for
+    # calculation of combined filters
+    if filters_combined is not None:
+        for x in filters_combined:
+            y = x[0]
+            z = x[1]
+            timeranges[str(y + "_" + z)] = timeranges[y].intersection(timeranges[z])
+
+    # Intersect datetimes of interest with the low-res tidal model
```
Comment (robbibt) [inline intertidal/exposure.py]:
```suggestion
    # Intersect datetimes of interest with the 1D tidal model
```

File: intertidal/elevation.py [hunk 6]
```diff
@@ -1264,22 +1264,24 @@ def intertidal_cli(
...
-            # Select times used for exposure modelling
-            all_times = pd.date_range(
-                start=round_date_strings(start_date, round_type="start"),
-                end=round_date_strings(end_date, round_type="end"),
-                freq=modelled_freq,
-            )
-
             # Calculate exposure
-            ds["exposure"], tide_cq = exposure(
+            exposure_filters, tide_cq_dict = exposure(
```
Comment (robbibt) [inline intertidal/elevation.py]: In the function below, `tide_cq_dict` is called `modelledtides_ds` - can we use that terminology here too for consistency? 

(are these raw modelled tides or tide quantiles? Alternatively, we could call it something like `tide_quantiles_ds` too if that was more clear)

Comment (erialC-P) [reply intertidal/elevation.py]: I'll update that variable name. It's a legacy hangover from an earlier version of the code.

**Review by robbibt — APPROVED** [review]
Wooo!

**Pull request merged**